rstar/src/algorithm/removal.rs

```diff
@@ -18,43 +18,70 @@ where
     Params: RTreeParams,
     R: SelectionFunction<T>,
 {
-    let mut result = None;
+    remove_recursive::<_, Params, _>(node, removal_function, false).pop()
+}
+
+pub fn remove_all<T, Params, R>(node: &mut ParentNode<T>, removal_function: &R) -> Vec<T>
+where
+    T: RTreeObject,
+    Params: RTreeParams,
+    R: SelectionFunction<T>,
+{
+    remove_recursive::<_, Params, _>(node, removal_function, true)
+}
+
+fn remove_recursive<T, Params, R>(
+    node: &mut ParentNode<T>,
+    removal_function: &R,
+    all_matches: bool,
+) -> Vec<T>
+where
+    T: RTreeObject,
+    Params: RTreeParams,
+    R: SelectionFunction<T>,
+{
+    let mut result = Vec::new();
     if removal_function.should_unpack_parent(&node.envelope) {
-        let mut removal_index = None;
-        for (index, child) in node.children.iter_mut().enumerate() {
+        let mut i = 0;
+        while i < node.children.len() {
+            let child = &mut node.children[i];
             match child {
                 RTreeNode::Parent(ref mut data) => {
-                    result = remove::<_, Params, _>(data, removal_function);
-                    if result.is_some() {
-                        if data.children.is_empty() {
-                            // Mark child for removal if it has become empty
-                            removal_index = Some(index);
+                    result.append(&mut remove_recursive::<_, Params, _>(
+                        data,
+                        removal_function,
+                        all_matches,
+                    ));
+                    if !result.is_empty() && data.children.is_empty() {
```

Comment:
hmm, I haven't looked at this algorithm that closely, but why did you combine these boolean expressions, it seems like there may now be cases where we won't hit the `break` - are any of them reachable?

Reply:
This checks whether we have removed any elements from the current subtree (-> `!result.is_empty()`) and if that is the case it tests whether that subtree has any elements left in it (-> `data.chlidren.is_empty()`). If not, it removes the current node since it is empty. 

Regarding `break`, you are correct. If we have found a result inside the current subtree but it is not empty after removing the match, then we do not cancel the search even though the flag is set.